### DISCUSSION TITLE
[BUG]

### DISCUSSION THREAD

**Discussion by ajaypsiborg** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1081] [Q&A]
I have used c++ producer sdk for raspberry pi for live video and audio streaming .I have installed Ubuntu os in raspberry pi4 model and i am successfully access the live video streaming with this command (./kvs_gstreamer_sample kvs_example_camera_stream) but while i want to access audio with video i am not able to access the live video and audio stream with this command and configured these command in .sh file.

Note =>but normal video stream working fine
kvs_gstreamer_audio_video_sample kvs_example_camera_stream
configured credential
export AWS_DEFAULT_REGION=ap-south-1
export AWS_ACCESS_KEY_ID= ***********************
export AWS_SECRET_ACCESS_KEY=*******************
export GST_PLUGIN_PATH=pwd/amazon-kinesis-video-streams-producer-sdk-cpp/build
export LD_LIBRARY_PATH=pwd/open-source/local/lib
export AWS_KVS_AUDIO_DEVICE=hw:2,0
cd amazon-kinesis-video-streams-producer-sdk-cpp/build/
./kvs_gstreamer_audio_video_sample kvs_example_camera_stream

getting error--------------------------------
bash: cd: amazon-kinesis-video-streams-producer-sdk-cpp/build/: No such file or directory
[INFO ] [03-10-2023 13:00:54:611.167 GMT] Using region: ap-south-1
[INFO ] [03-10-2023 13:00:54:612.026 GMT] Using aws credentials for Kinesis Video Streams
[INFO ] [03-10-2023 13:00:54:612.382 GMT] No session token was detected.
[INFO ] [03-10-2023 13:00:54:616.691 GMT] createKinesisVideoClient(): Creating Kinesis Video Client
[WARN ] [03-10-2023 13:00:54:616.977 GMT] fixupClientInfo(): Connection timeout is invalid...setting to default
[WARN ] [03-10-2023 13:00:54:617.105 GMT] fixupClientInfo(): Completion timeout is invalid...setting to default
[INFO ] [03-10-2023 13:00:54:617.245 GMT] heapInitialize(): Initializing native heap with limit size 134217728, spill ratio 0% and flags 0x00000001
[INFO ] [03-10-2023 13:00:54:617.345 GMT] heapInitialize(): Creating AIV heap.
[INFO ] [03-10-2023 13:00:54:617.519 GMT] heapInitialize(): Heap is initialized OK
[DEBUG] [03-10-2023 13:00:54:618.034 GMT] getSecurityTokenHandler invoked
[DEBUG] [03-10-2023 13:00:54:618.356 GMT] Refreshing credentials. Force refreshing: 0 Now time is: 1696338054618221939 Expiration: 0
[INFO ] [03-10-2023 13:00:54:618.561 GMT] New credentials expiration is 1696341654
[INFO ] [03-10-2023 13:00:54:618.750 GMT] createDeviceResultEvent(): Create device result event.
[DEBUG] [03-10-2023 13:00:54:618.902 GMT] clientReadyHandler invoked
[DEBUG] [03-10-2023 13:00:54:619.135 GMT] Client is ready
[INFO ] [03-10-2023 13:00:54:619.585 GMT] Creating Kinesis Video Stream kvs_example_camera_stream
[INFO ] [03-10-2023 13:00:54:619.920 GMT] createKinesisVideoStream(): Creating Kinesis Video Stream.
[INFO ] [03-10-2023 13:00:54:620.061 GMT] logStreamInfo(): SDK version: https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/commit/de4299f942bb89c867d0dc73aeb4bc875d560a8f
[DEBUG] [03-10-2023 13:00:54:620.146 GMT] logStreamInfo(): Kinesis Video Stream Info
[DEBUG] [03-10-2023 13:00:54:620.216 GMT] logStreamInfo(): Stream name: kvs_example_camera_stream
[DEBUG] [03-10-2023 13:00:54:620.302 GMT] logStreamInfo(): Streaming type: STREAMING_TYPE_REALTIME
[DEBUG] [03-10-2023 13:00:54:620.394 GMT] logStreamInfo(): Content type: video/h264,audio/aac
[DEBUG] [03-10-2023 13:00:54:620.484 GMT] logStreamInfo(): Max latency (100ns): 600000000
[DEBUG] [03-10-2023 13:00:54:620.573 GMT] logStreamInfo(): Fragment duration (100ns): 20000000
[DEBUG] [03-10-2023 13:00:54:620.668 GMT] logStreamInfo(): Key frame fragmentation: Yes
[DEBUG] [03-10-2023 13:00:54:620.756 GMT] logStreamInfo(): Use frame timecodes: Yes
[DEBUG] [03-10-2023 13:00:54:620.843 GMT] logStreamInfo(): Absolute frame timecodes: Yes
[DEBUG] [03-10-2023 13:00:54:620.930 GMT] logStreamInfo(): Nal adaptation flags: 0
[DEBUG] [03-10-2023 13:00:54:621.026 GMT] logStreamInfo(): Average bandwidth (bps): 4194304
[DEBUG] [03-10-2023 13:00:54:621.114 GMT] logStreamInfo(): Framerate: 100
[DEBUG] [03-10-2023 13:00:54:621.204 GMT] logStreamInfo(): Buffer duration (100ns): 1200000000
[DEBUG] [03-10-2023 13:00:54:621.292 GMT] logStreamInfo(): Replay duration (100ns): 400000000
[DEBUG] [03-10-2023 13:00:54:621.380 GMT] logStreamInfo(): Connection Staleness duration (100ns): 600000000
[DEBUG] [03-10-2023 13:00:54:621.474 GMT] logStreamInfo(): Store Pressure Policy: 1
[DEBUG] [03-10-2023 13:00:54:621.566 GMT] logStreamInfo(): View Overflow Policy: 1
[DEBUG] [03-10-2023 13:00:54:621.656 GMT] logStreamInfo(): Allow stream creation: Yes
[DEBUG] [03-10-2023 13:00:54:621.749 GMT] logStreamInfo(): Segment UUID: NULL
[DEBUG] [03-10-2023 13:00:54:621.841 GMT] logStreamInfo(): Frame ordering mode: 5
[DEBUG] [03-10-2023 13:00:54:621.938 GMT] logStreamInfo(): Track list
[DEBUG] [03-10-2023 13:00:54:622.009 GMT] logStreamInfo(): Track id: 1
[DEBUG] [03-10-2023 13:00:54:622.102 GMT] logStreamInfo(): Track name: kinesis_video
[DEBUG] [03-10-2023 13:00:54:622.189 GMT] logStreamInfo(): Codec id: V_MPEG4/ISO/AVC
[DEBUG] [03-10-2023 13:00:54:622.232 GMT] logStreamInfo(): Track type: TRACK_INFO_TYPE_VIDEO
[DEBUG] [03-10-2023 13:00:54:622.306 GMT] logStreamInfo(): Track cpd: NULL
[DEBUG] [03-10-2023 13:00:54:622.407 GMT] logStreamInfo(): Track id: 2
[DEBUG] [03-10-2023 13:00:54:622.506 GMT] logStreamInfo(): Track name: audio
[DEBUG] [03-10-2023 13:00:54:622.702 GMT] logStreamInfo(): Codec id: A_AAC
[DEBUG] [03-10-2023 13:00:54:622.800 GMT] logStreamInfo(): Track type: TRACK_INFO_TYPE_AUDIO
[DEBUG] [03-10-2023 13:00:54:622.894 GMT] logStreamInfo(): Track cpd: NULL
[INFO ] [03-10-2023 13:00:54:980.613 GMT] writeHeaderCallback(): RequestId: 98d9ebac-313f-437c-9f89-cda98c6b6225
[DEBUG] [03-10-2023 13:00:54:980.912 GMT] describeStreamCurlHandler(): [] DescribeStream API response: {"StreamInfo":{"CreationTime":1.696324714545E9,"DataRetentionInHours":24,"DeviceName":"Kinesis_Video_Device","IngestionConfiguration":null,"KmsKeyId":"arn:aws:kms:ap-south-1:550053710858:alias/aws/kinesisvideo","MediaType":"video/h264","Status":"ACTIVE","StreamARN":"arn:aws:kinesisvideo:ap-south-1:550053710858:stream/kvs_example_camera_stream/1696324714545","StreamName":"kvs_example_camera_stream","Version":"BGHMLuQTPyTxnQKrFcFQ"}}
[INFO ] [03-10-2023 13:00:54:984.225 GMT] describeStreamResultEvent(): Describe stream result event.
[WARN ] [03-10-2023 13:00:54:984.423 GMT] describeStreamResult(): Retention period returned from the DescribeStream call doesn't match the one specified in the StreamInfo
[WARN ] [03-10-2023 13:00:54:984.489 GMT] describeStreamResult(): Content type returned from the DescribeStream(video/h264) call doesn't match the one specified in the StreamInfo(video/h264,audio/aac)
[WARN ] [03-10-2023 13:00:54:984.539 GMT] describeStreamResult(): [kvs_example_camera_stream] Retention period returned from the DescribeStream call doesn't match the one specified in the StreamInfo
[WARN ] [03-10-2023 13:00:54:984.587 GMT] describeStreamResult(): [kvs_example_camera_stream] Content type returned from the DescribeStream call doesn't match the one specified in the StreamInfo
[INFO ] [03-10-2023 13:00:55:338.168 GMT] writeHeaderCallback(): RequestId: a2311647-6a2b-4505-a4d0-e08d33884326
[DEBUG] [03-10-2023 13:00:55:338.556 GMT] getStreamingEndpointCurlHandler(): [kvs_example_camera_stream] GetStreamingEndpoint API response: {"DataEndpoint":"[https://s-703775db.kinesisvideo.ap-south-1.amazonaws.com"}](https://s-703775db.kinesisvideo.ap-south-1.amazonaws.com%22%7D/)
[INFO ] [03-10-2023 13:00:55:344.436 GMT] getStreamingEndpointResultEvent(): Get streaming endpoint result event.
[DEBUG] [03-10-2023 13:00:55:344.805 GMT] getStreamingTokenHandler invoked
[DEBUG] [03-10-2023 13:00:55:344.957 GMT] Refreshing credentials. Force refreshing: 1 Now time is: 1696338055344935848 Expiration: 1696341654
[INFO ] [03-10-2023 13:00:55:345.076 GMT] New credentials expiration is 1696341655
[INFO ] [03-10-2023 13:00:55:345.217 GMT] getStreamingTokenResultEvent(): Get streaming token result event.
[DEBUG] [03-10-2023 13:00:55:345.385 GMT] streamReadyHandler invoked
[DEBUG] [03-10-2023 13:00:55:345.999 GMT] Stream is ready: kvs_example_camera_stream
[DEBUG] [03-10-2023 13:00:55:474.421 GMT] appsink configured
[INFO ] [03-10-2023 13:00:55:598.252 GMT] Using default video device
[INFO ] [03-10-2023 13:00:55:602.745 GMT] Using audio device hw:2,0
Not all elements could be created:
[INFO ] [03-10-2023 13:00:55:606.321 GMT] stopKinesisVideoStreamSync(): Synchronously stopping Kinesis Video Stream 0000aaaaeb434280.
[DEBUG] [03-10-2023 13:00:55:606.451 GMT] streamClosedHandler invoked for upload handle: 18446744073709551615
[INFO ] [03-10-2023 13:00:55:606.527 GMT] Freeing Kinesis Video Stream for kvs_example_camera_stream
[INFO ] [03-10-2023 13:00:55:606.588 GMT] freeKinesisVideoStream(): Freeing Kinesis Video stream.
[DEBUG] [03-10-2023 13:00:55:606.628 GMT] freeStream(): [kvs_example_camera_stream] Freeing stream
[DEBUG] [03-10-2023 13:00:55:606.669 GMT] curlApiCallbacksShutdownActiveRequests(): pActiveRequests hashtable is empty
[DEBUG] [03-10-2023 13:00:56:007.183 GMT] freeStream(): [kvs_example_camera_stream] Completed freeing stream
[INFO ] [03-10-2023 13:00:56:007.661 GMT] Freeing Kinesis Video Stream for kvs_example_camera_stream
[INFO ] [03-10-2023 13:00:56:007.892 GMT] freeKinesisVideoClient(): Freeing Kinesis Video Client
[DEBUG] [03-10-2023 13:00:56:008.020 GMT] curlApiCallbacksShutdownActiveRequests(): pActiveRequests hashtable is empty
[DEBUG] [03-10-2023 13:00:56:408.752 GMT] freeKinesisVideoClientInternal(): Total allocated memory 0
[INFO ] [03-10-2023 13:00:56:409.092 GMT] Completed freeing client
[WARN ] [03-10-2023 13:00:56:409.286 GMT] curlApiCallbacksShutdown(): curlApiCallbacksShutdown called when already in progress of shutting down

also applied the second method but it will also does not work so kindy suggest with above method
export GST_PLUGIN_PATH=pwd/build
export LD_LIBRARY_PATH=pwd/open-source/local/lib

gst-launch-1.0 -v v4l2src device=/dev/video0 ! videoconvert ! x264enc speed-preset=ultrafast tune=zerolatency ! h264parse ! kvssink stream-name="kvs_example_camera_stream" access-key="****************" secret-key="************************** alsasrc device=hw:2,0 format=S16LE rate=44100 ! audioconvert ! voaacenc ! queue ! kvssink

**Comment by niyatim23** [comment]
Hi @ajaypsiborg, can you try using `appsink` or `filesink` instead of `kvssink` and see if you can access the camera / mic of your device? This will help us identify if the issue is with the device / pipeline or with `kvssink`. Also, can you try running your sh file with `sudo`
